### PR TITLE
Upgrade golang to v1.16

### PR DESCRIPTION
**Pull request by mctofu**
This allows use of the [new embed package](https://golang.org/doc/go1.16#library-embed) added in 1.16.

There's also a few [notable updates](https://golang.org/doc/go1.16#go-command) for modules:
> The go mod vendor and go mod tidy subcommands now accept the -e flag, which instructs them to proceed despite errors in resolving missing packages. ([noted here](https://github.com/dependabot/dependabot-core/blob/f3239dfae253e916b62f9b4b167eeceb6b3ba25a/go_modules/lib/dependabot/go_modules/file_updater/go_mod_updater.rb#L131))

> Module-aware mode is enabled by default, regardless of whether a go.mod file is present in the current working directory or a parent directory. More precisely, the GO111MODULE environment variable now defaults to on. 

We could omit `GO111MODULE=on` from the helper scripts.

### CODE REVIEW

**Review by xlgmokha — APPROVED** [review]

**Comment by xlgmokha** [comment]
https://github.com/dependabot/dependabot-core/pull/3235 should help fix the failing tests.

**Comment by mctofu** [comment]
The tests are showing a behavior change when there is a reference to a package that has been renamed. In [the test](https://github.com/dependabot/dependabot-core/blob/f3239dfae253e916b62f9b4b167eeceb6b3ba25a/go_modules/spec/dependabot/go_modules/file_updater/go_mod_updater_spec.rb#L226) we are expecting that `go mod tidy` would complain about this but we are now seeing it fail prior to that at the `go get -d` step.
```
$ go get -d
github.com/dependabot/vgotest imports
        github.com/googleapis/gnostic/OpenAPIv2: cannot find module providing package github.com/googleapis/gnostic/OpenAPIv2
```

**Comment by lovemybabygirl42000** [comment]
Let me say this zack backer pat and Kevin high an you old bitches in town
big Oakland cop won't do anything I will an am and already have don't crank
your trucks tonight

**Review by thepwagner — APPROVED** [review]
🎉 
Aside: https://github.com/thepwagner/action-update-dockerurl supports the `Dockerfile` syntax we're using for this dependency, and https://github.com/thepwagner/action-update-brewformula/blob/main/brew/golang.go is gnarly, but handles parsing https://golang.org/dl/?mode=json - we could automate these bumps!

I'd dig some tests that cover the new `retract` directive too, e.g. to ensure Dependabot won't propose an update to a retracted version. As noted, there are a lot of module goodies in `1.16`.

**Comment by mctofu** [comment]
I pushed an update to fix the [failing test](https://github.com/dependabot/dependabot-core/blob/f3239dfae253e916b62f9b4b167eeceb6b3ba25a/go_modules/spec/dependabot/go_modules/file_updater/go_mod_updater_spec.rb#L226) that is worth re-reviewing: https://github.com/dependabot/dependabot-core/pull/3233/commits/f94051a9d83cabf8e76bb4a55b891fc30058033d.

The issue is that `go get -d` now fails if the code references a package that can't be found so I've reverted to our previous behavior of raising `DependencyFileNotResolvable` (https://github.com/dependabot/dependabot-core/pull/2627). However, since the new error message doesn't differentiate between the module not existing vs existing but not providing the package I also had to make some adjustments to prevent the error from raising `GitDependenciesNotReachable` instead. I found that running `go list -m -versions #{repo_path}` was a workable way to test that a module's repo was reachable and was potentially lighter weight than using `go get` since it doesn't download the module's dependencies.

**Comment by mctofu** [comment]
> I'd dig some tests that cover the new `retract` directive too, e.g. to ensure Dependabot won't propose an update to a retracted version.

@thepwagner: I added a test that we won't strip the `retract` directive from a `go.mod`. Avoiding retracted versions isn't working yet. I think we need to update the extracted go module code to get the latest behavior updates there.

**Review by thepwagner — APPROVED** [review]
> I think we need to update the extracted go module code to get the latest behavior updates there.

I assumed since [0.4.1](https://github.com/dependabot/dependabot-core/blob/main/go_modules/helpers/go.mod#L9) included https://github.com/golang/mod/commit/c0d644d00ab849f4506f17a98a5740bf0feff020#diff-320fe6a1bc7a6321d706e6c7fc95d457cd5ea1b13425b320d56897c78074b350 , we'd have the necessary bits to update the helpers.

File: Dockerfile [hunk 1]
```diff
@@ -137,8 +137,8 @@ RUN add-apt-repository ppa:ondrej/php \
 ### GO
 
 # Install Go and dep
-ARG GOLANG_VERSION=1.15.7
-ARG GOLANG_CHECKSUM=0d142143794721bb63ce6c8a6180c4062bcf8ef4715e7d6d6609f3a8282629b3
+ARG GOLANG_VERSION=1.16
+ARG GOLANG_CHECKSUM=013a489ebb3e24ef3d915abe5b94c3286c070dfe0818d5bca8108f1d6e8440d2
```
[the remark this answers is not in the document]
Comment (mctofu) [reply Dockerfile]: Thanks! I bumped this to 1.16.2 which was released shortly afterwards.

**Comment by mctofu** [comment]
> I assumed since [0.4.1](https://github.com/dependabot/dependabot-core/blob/main/go_modules/helpers/go.mod#L9) included [golang/mod@c0d644d#diff-320fe6a1bc7a6321d706e6c7fc95d457cd5ea1b13425b320d56897c78074b350](https://github.com/golang/mod/commit/c0d644d00ab849f4506f17a98a5740bf0feff020#diff-320fe6a1bc7a6321d706e6c7fc95d457cd5ea1b13425b320d56897c78074b350) , we'd have the necessary bits to update the helpers.

Aha! I had been hoping that `modfetch.Lookup("direct", args.Dependency.Name)` would have already filtered retracted versions in 1.16. That doesn't seem to be the case so we will need to filter on our end.

**Comment by lovemybabygirl42000** [comment]
Try it again and I'll .show the world want to see Nancy

Click this link: https://www.houzz.com/writeReview/cmd=r/n=byrd_byrd56/lclck=1/e=TqhUFri081Zwqq8v8n8NJLD3us7F9qfM_G__4cD2XXBHSFlQVF3RAC3TIf7XBE1zw7w-tEDPqPMDYUrdWDp8ZyreT9YVJw7V/v=b to write your review. 


Thanks in advance and let me know if you have any questions,

ByrdandJess antiques

**Pull request merged**